File: go/problem_21.go
package main

import "fmt"

var cache = make(map[int]int)

func d(n int) int {
	var sum, ok = cache[n]
	if !ok {
		sum = 1
		for _, v := range Factors(n) {
			sum += v
		}
		cache[n] = sum
	}
	return sum
}

func main() {
	sum := 0
	for i := 0; i < 10000; i++ {
		if i == d(d(i)) && i != d(i) {
			sum += i
		}
	}
	fmt.Printf("%d\n", sum)
}

File: dart/problem_21.dart
#!/usr/bin/env dart

import 'dart:collection';
import 'factor.dart';
import 'iterables.dart';
import 'package:collection/collection.dart';

var cache = HashMap<int, int>();
int d(int n) => cache.putIfAbsent(n, () => 1 + factor(n).sum);

void main() {
  print(range(1, 9999).where((x) => x == d(d(x)) && x != d(x)).sum);
}
